addons/hw_drivers/controllers/driver.py

```diff
@@ -560,19 +560,27 @@ def run(self):
 
 class SocketManager(Thread):
 
+    def __init__(self):
+        super(SocketManager, self).__init__()
+        self.open_socket(9000)
+
+    def open_socket(self, port):
+        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self.sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        self.sock.bind(('', port))
+        self.sock.listen(1)
+        self.sock.setblocking(False)
+
     def run(self):
         while True:
             try:
-                sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-                sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
-                sock.bind(('', 9000))
-                sock.listen(1)
-                dev, addr = sock.accept()
+                dev, addr = self.sock.accept()
                 if addr and addr[0] not in socket_devices:
                     iot_device = IoTDevice(type('', (), {'dev': dev}), 'socket')
                     socket_devices[addr[0]] = iot_device
             except OSError as e:
-                _logger.error(_('Error in SocketManager: %s') % (e.strerror))
+                pass
+            time.sleep(30)
```

Comment:
`Setblocking(False)` is exactly the fact that you can add more than 1 terminal.
if it is `True` the run is blocked with `self.sock` which is not connected to the terminal. 
He's waiting for a connection he'll never have.
With `False` the socket is not blocked and will listen again after x seconds
Another solution would have been to put a timeout on the `self.sock` object
The 30 seconds are arbitrary. More or less why

Reply:
Since the SocketManager has its own thread, why is the `run` blocking on the `accept` an issue? If two terminals connect, `accept` will return a new socket twice, which gets put into `socket_devices`, where the IngenicoDrivers will read from and write to them. If a new terminal connects or one of the existing ones restarts the connection, `accept` will return again and update `socket_devices` if needed.

If two terminals connect at the exact same time, the `listen(1)` might have resulted in the second one getting rejected at first, because only 1 connection was allowed to be queued.

Reply:
Not exactly.
First step the terminal is connected.
When the terminal is disconnected he don't release the blocking port.
And when he want to reconnect to the box the port is blocked so he can't enable a new connection.

But if we don't block the port, after 30 seconds, the thread listen again for a new connection and we can re-connect the terminal or another one. When we use 2 terminal there are a delay between the detection of the second.

The backlog is not `the allowed connection` but `the number of unaccepted connections`.
The proof is that we can manage more than 1 terminal with `socket.listen(1)`

With your reflexion there again a problem.
We can put more than 1 connections but it is a `while True` so we hit the default number after x loop of unaccepted connections.


There are 2 terminals at the office you can try another way.
But the terminals must stay at the IOT office

Reply:
> The backlog is not the allowed connection but the number of unaccepted connections.

Indeed, I've understood it to be the amount of connections that can be queued up while waiting to be accepted.

> We can put more than 1 connections but it is a while True so we hit the default number after x loop of unaccepted connections.

What do you understand to be an unaccepted connection? What does looping over unaccepted connections mean?

> When the terminal is disconnected he don't release the blocking port.

I don't understand what you mean here with the blocking port and it not being released. I'll give a code example.

# Server
```python
import socket
import threading

def f(name, dev):
	print("Hello {}".format(name))
	dev.recv(4)
	print("Bye {}".format(name))

sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
sock.bind(('', 9000))
sock.listen(1)
print(sock)

sock1, addr1 = sock.accept()
print(sock1)
t1 = threading.Thread(target=f, args=('1', sock1))
t1.start()

sock2, addr2 = sock.accept()
print(sock2)
t2 = threading.Thread(target=f, args=('2', sock2))
t2.start()

sock3, addr3 = sock.accept()
print(sock3)
t3 = threading.Thread(target=f, args=('3', sock3))
t3.start()
```

# Client
```python
import socket
import time

s1 = socket.socket()
# Replace the IP with the one the server is running on.
s1.connect(('192.168.1.4', 9000))
time.sleep(2)
s2 = socket.socket()
s2.connect(('192.168.1.4', 9000))
time.sleep(2)
s1.close()
time.sleep(2)
s3 = socket.socket()
s3.connect(('192.168.1.4', 9000))
```

Run the server code in one Python shell instance and the client code after that in a different one.

In my case the output I get on the server side is
```
<socket.socket fd=15, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('0.0.0.0', 9000)>
<socket.socket fd=16, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('192.168.1.4', 9000), raddr=('192.168.1.4', 48296)>
Hello 1
<socket.socket fd=17, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('192.168.1.4', 9000), raddr=('192.168.1.4', 48298)>
Hello 2
Bye 1
<socket.socket fd=18, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('192.168.1.4', 9000), raddr=('192.168.1.4', 48300)>
Hello 3
```

The first socket with `laddr=('0.0.0.0', 9000)` and no `raddr` is the one corresponding to `self.sock` in SocketManager. The other 3 were created by `accept` with a corresponding `raddr` set, and would correspond to the ones that get stored in `socket_devices`. In what way is a terminal blocked from connecting in this scenario?

Reply:
Ok now put your `sock.accept()` in a `while True` loop and test to add and remove connection.
But your example is the same as the old version.

I'm not against a refactoring of Ingenico if it can improve stability. 
We can work together in this if you want at the office the next week.

Reply:
I'm not sure if the assumptions made in this PR about socket behavior are correct. Since we're dealing with hard to reproduce issues, it's useful to verify that. 30 seconds is a long time to not accept any new connections. Could you answer the questions from my last comment, in particular what you understand to be an unaccepted connection, what you mean with looping over unaccepted connections, and why you think terminals can get blocked from making new connections when the `self.sock` socket is in blocking mode and `accept` is called in a `while True` loop while multiple terminals connect and disconnect?

It's not so much about refactoring as it's about making sure the new code (which has currently lead to greater instability for the customers) does what we think it does.